hazelcast/src/main/java/com/hazelcast/client/impl/protocol/ClientMessage.java

```diff
@@ -16,367 +16,110 @@
-
-    /**
-     * End Flag
-     */
-    public static final short END_FLAG = 0x40;
-
-    /**
-     * Begin and End Flags
-     */
-    public static final short BEGIN_AND_END_FLAGS = (short) (BEGIN_FLAG | END_FLAG);
-
-    /**
-     * Listener Event Flag
-     */
-    public static final short LISTENER_EVENT_FLAG = 0x01;
-
-    /**
-     * ClientMessage Fixed Header size in bytes
-     */
-    public static final int HEADER_SIZE;
-
-    private static final String PROP_HAZELCAST_PROTOCOL_UNSAFE = "hazelcast.protocol.unsafe.enabled";
-    private static final boolean USE_UNSAFE = Boolean.getBoolean(PROP_HAZELCAST_PROTOCOL_UNSAFE);
-
-    private static final int FRAME_LENGTH_FIELD_OFFSET = 0;
-    private static final int VERSION_FIELD_OFFSET = FRAME_LENGTH_FIELD_OFFSET + Bits.INT_SIZE_IN_BYTES;
-    private static final int FLAGS_FIELD_OFFSET = VERSION_FIELD_OFFSET + Bits.BYTE_SIZE_IN_BYTES;
-    private static final int TYPE_FIELD_OFFSET = FLAGS_FIELD_OFFSET + Bits.BYTE_SIZE_IN_BYTES;
-    private static final int CORRELATION_ID_FIELD_OFFSET = TYPE_FIELD_OFFSET + Bits.SHORT_SIZE_IN_BYTES;
-    private static final int PARTITION_ID_FIELD_OFFSET = CORRELATION_ID_FIELD_OFFSET + Bits.LONG_SIZE_IN_BYTES;
-    private static final int DATA_OFFSET_FIELD_OFFSET = PARTITION_ID_FIELD_OFFSET + Bits.INT_SIZE_IN_BYTES;
-
-    static {
-
-        HEADER_SIZE = DATA_OFFSET_FIELD_OFFSET + Bits.SHORT_SIZE_IN_BYTES;
-    }
-
-    private transient int writeOffset;
+@SuppressWarnings("checkstyle:MagicNumber")
+public final class ClientMessage implements OutboundFrame, Iterable<ClientMessage.Frame> {
+
+    public static final int TYPE_FIELD_OFFSET = 0;
+    public static final int CORRELATION_ID_FIELD_OFFSET = TYPE_FIELD_OFFSET + Bits.SHORT_SIZE_IN_BYTES;
+
+    //offset valid for fragmentation frames only
+    public static final int FRAGMENTATION_ID_OFFSET = 0;
+
+    //optional fixed partition id field offset
+    public static final int PARTITION_ID_FIELD_OFFSET = CORRELATION_ID_FIELD_OFFSET + Bits.LONG_SIZE_IN_BYTES;
+
+    public static final int DEFAULT_FLAGS = 0;
+    public static final int BEGIN_FRAGMENT = 1 << 15;
+    public static final int END_FRAGMENT = 1 << 14;
+    public static final int UNFRAGMENTED_MESSAGE = BEGIN_FRAGMENT | END_FRAGMENT;
+    public static final int FINAL = 1 << 13;
+    public static final int BEGIN_DATA_STRUCTURE = 1 << 12;
+    public static final int END_DATA_STRUCTURE = 1 << 11;
+    public static final int IS_NULL = 1 << 10;
+    public static final int IS_EVENT = 1 << 9;
+
+    //frame length + flags
+    public static final int SIZE_OF_FRAMELENGHT_AND_FLAGS = Bits.INT_SIZE_IN_BYTES + Bits.SHORT_SIZE_IN_BYTES;
+    public static final Frame NULL_FRAME = new Frame(null, IS_NULL);
+    public static final Frame BEGIN_FRAME = new Frame(null, BEGIN_DATA_STRUCTURE);
+    public static final Frame END_FRAME = new Frame(null, END_DATA_STRUCTURE);
+
+    private LinkedList<Frame> frames;
```

Comment:
Would we consider turning the ClientMessage into a linked list itself? We could do it by replacing the `LinkedList` with a field `.Next` in the `Frame` itself and store in the ClientMessage:

```java
private Frame head
private Frame tail
```

Whenever a frame is added, it would 
```java
this.tail.Next = frame;
this.tail = frame;
```

The reasons for this:

- one less hop for accessing frames
- whenever a frame is accessed directly, it's head (frames.get(0)), this potentially could be faster
- less allocations (however the linked list is built)

As an additional cost, the iterator would need to be implemented for the `getFrames`

Reply:
This could be a good improvement. We will have a couple of performance-related improvements. We will prepare them right after this one. Your suggestion could a different pr. I will work on the idea. One alternative is to make ClientMEssage extend LinkedList. It might be a lot easier

Reply:
@Scooletz I have made client message extend LinkedList.

Reply:
Do you have any performance benchmark for showing one is better than the other in access time? I recommend doing a JMH test to see the difference if indeed improves as suggested above.

Reply:
@ihsandemir If that's me who's being asked I do not have these. I'm not sure that I can provide them in JMH within a reasonable timeframe (I don't know know the ecosystem), but for sure I could do in on .NET side with the `Benchmarkdotnet` tooling when doing .NET client port.

Just to clarify: this was a remark, not a request change 😛

Reply:
Code also became much simpler, even if this does not give any performance improvement, it is still a good change.